docs/sources/developers/plugins/working-with-data-frames.md

```diff
@@ -0,0 +1,109 @@
+
+The [DataFrame]({{< relref "../../packages_api/data/dataframe.md" >}}>) interface contains a `name` and an array of `fields` where each field contains the name, type, and the values for the field.
+
+## Create a data frame
+
+If you building a data source plugin, you'll most likely want to convert a response from an external API to a data frame. Let's look at how to create a data frame.
+
+Let's start with creating a simple data frame that represents a time series. The easiest way to create a data frame is to use the [toDataFrame]({{< relref "../../packages_api/data/todataframe.md" >}}) function.
+
+```ts
+// Need to be of the same length.
+const timeValues = [1599471973065, 1599471975729];
+const numberValues = [12.3, 28.6];
+
+// Create data frame from values.
+const frame = toDataFrame({
+    name: "http_requests_total",
+    fields: [
+        { name: "Time", type: FieldType.time, values: timeValues },
+        { name: "Value", type: FieldType.number, values: numberValues }
+    ]
+});
+```
+
+> **Note:** Data frames representing time series contain at least a `time` field, and a `number` field. By convention, built-in plugins use `Time` and `Value` as field names for data frames containing time series data.
+
+As you can see from the example, creating data frames like this requires that your data is already stored as columnar data. If you already have the records in the form of an array of objects, you can pass it to `toDataFrame` which tries to guess the schema based on the types and names of the objects in the array. If you're creating complex data frames this way, make sure to verify that you get the schema you expect.
+
+```ts
+const series = [
+    { Time: 1599471973065, Value: 12.3 },
+    { Time: 1599471975729, Value: 28.6 }
+];
+
+const frame = toDataFrame(series);
+frame.name = 'http_requests_total'
+```
+
+## Read values from a data frame
+
+When you're building a panel plugin, the data frames returned by the data source are available from the `data` prop in your panel component.
+
+```ts
+const SimplePanel: React.FC<Props> = ({ data }) => {
+  const frame = data.series[0];
+
+  // ...
+}
+```
+
+Before you start reading the data, think about what data you expect. For time series, we can expect a `Time` and a `Value` field, so it's safe to hard code them.
+
+```ts
+const timeField = frame.fields.find(field => field.name === 'Time');
+if (timeField.type !== FieldType.time) {
+  throw new Error('Invalid type for Time field')
+}
+
+const valueField = frame.fields.find(field => field.name === 'Value');
+if (valueField.type !== FieldType.number) {
+  throw new Error('Invalid type for Value field')
+}
+```
+
+Consider other types of visualizations may need multiple dimensions. For example, a bubble chart that uses three numeric fields: The X-axis, Y-axis, and one for the radius of each bubble. In this case, instead of hard coding the field names, we recommend that you let the user choose the field to use for each dimension.
```

Comment:
```suggestion
Other types of visualizations might need multiple dimensions. For example, a bubble chart that uses three numeric fields: the X-axis, Y-axis, and one for the radius of each bubble. In this case, instead of hard coding the field names, we recommend that you let the user choose the field to use for each dimension.
```